### PR TITLE
SDK-886 `getDefaultUriScheme breaks if URI scheme used for Branch is not first`

### PR DESCRIPTION
**Pull request by illphil610**
This PR is a solution to SDK-886 where the validation fails if the URI scheme used for Branch is not the first in the plist file.  I did notice that `getDefaultUriScheme()` is used in more areas than just in the validation logic which is why I decided to create a new function to keep it separate to not interfere with internal logic.  I think since `getDefaultUriScheme()` only checks for the 1st entry we should refactor that to use the validated URI scheme that matches the server side URI scheme (if it exists) but didn't want to make that decision in this PR.  Just some thoughts!

Unit Testing Note:
I am having issues running unit tests locally.  I have unit tests created for the new methods but whenever I try to run them I am unable to build successful to run the tests thus I am not sure if they working correctly.  The tests are not included in this PR, but will be after I figure out the issue.  If we could have a working session to possibly solve my problem faster, I'd appreciate it!  Just wanted to give you a look into the solution in case I went down the wrong path to save time.

### CODE REVIEW

File: Branch-SDK/Branch-SDK/Branch+Validator.m [hunk 1]
```diff
@@ -28,6 +28,11 @@ static inline void BNCAfterSecondsPerformBlockOnMainThread(NSTimeInterval second
     dispatch_after(BNCDispatchTimeFromSeconds(seconds), dispatch_get_main_queue(), block);
 }
 
+typedef struct {
+    NSString *uriScheme;
+    BOOL doesMatchServerScheme;
+} ValidatedClientUriSchemeModel;
```
Comment (echo-branch) [inline Branch-SDK/Branch-SDK/Branch+Validator.m]: Prefix the struct definition with BNC.

File: Branch-SDK/Branch-SDK/BNCSystemObserver.h [hunk 2]
```diff
@@ -19,6 +19,7 @@
                           andType:(NSString *__autoreleasing*)type;
 + (NSString *)getVendorId;
 + (NSString *)getDefaultUriScheme;
++ (NSArray *) getAllClientUriSchemes;
```
Comment (echo-branch) [inline Branch-SDK/Branch-SDK/BNCSystemObserver.h]: If possible, let's type our collections in new code.

File: Branch-SDK/Branch-SDK/Branch+Validator.m [hunk 3]
```diff
@@ -55,6 +60,29 @@ - (void) startValidation {
...
+ This method will loop through the clients URI schemes and compare each index with the uri scheme in Branch to look for a match.
+ @param clientUriSchemes NSArray representing each URI scheme present in the users plist file
+ @param serverUriScheme NSString representing the URI scheme Branch has for the client
+ @returns ValidatedClientUriSchemeModel : Struct which contains the necessary info to toggle validation logic in testing
+ */
+- (ValidatedClientUriSchemeModel) validateClientScheme:(NSArray*)clientUriSchemes withServerScheme:(NSString*)serverUriScheme {
+    ValidatedClientUriSchemeModel model;
+    model.doesMatchServerScheme = NO;
+    
+    for (id uriScheme in clientUriSchemes) {
```
Comment (echo-branch) [inline Branch-SDK/Branch-SDK/Branch+Validator.m]: Let's avoid id and use typed collections whenever possible.

**Comment by echo-branch** [comment]
Let's add some unit tests around this.  I saw a dictionary length crash when testing in my test app.  Usually this is cause a dictionary is used instead of a string.  Didn't dig into it too much, but the error doesn't occur on stage.

File: Branch-SDK/Branch-SDK/Branch+Validator.m (hunk 1 above)
Comment (E-B-Smith) [reply Branch-SDK/Branch-SDK/Branch+Validator.m]: Also can't be a struct: use an object instead.

Structs can't hold obj-c objects.

Comment (echo-branch) [reply Branch-SDK/Branch-SDK/Branch+Validator.m]: That should have triggered an error. :/

**Pull request closed without merging**